### ISSUE TITLE
Codable metadata serializer?

### ISSUE THREAD

**Issue by noremac** [yapstudios/YapDatabase#503]
In the 4.0 release, it's nice and easy to set up a `Codable` serialization for objects:

``` swift
db.registerCodableSerialization(
  Article.self,
  forCollection: "Article"
)
```

However, it's a bit cumbersome to set one up for metadata:

``` swift
db.registerMetadataSerializer(
  YapDatabase.codableSerializer(Article.Metadata.self),
  forCollection: "Article"
)
db.registerMetadataDeserializer(
  YapDatabase.codableDeserializer(Article.Metadata.self),
  forCollection: "Article"
)
```

Am I missing anything here? I'd be happy to create a PR if there's interest, otherwise I can stick to my own local extension.

**Comment by robbiehanson** [comment]
It looks like `registerCodableSerialization` registers `Article.self` for BOTH objects & metadata. Perhaps this is unusual, and undesired behavior?

We could change the methods to be:
- registerCodableObjectSerialization
- registerCodableMetadataSerialization

What do you think?

**Comment by noremac** [comment]
> It looks like registerCodableSerialization registers Article.self for BOTH objects & metadata. Perhaps this is unusual, and undesired behavior?

Oh, I hadn't caught that! Yeah, I would think it would be rare that you'd have the same type for both the object and metadata?

I think my ideal signature would be something like:

``` swift
public func registerCodableSerialization<O, M>(object objectType: O.Type, metadata metadataType: M.Type, forCollection collection: String?) where O: Codable, M: Codable
```

This would be in addition to the existing method that only takes the "object" type.

What you proposed is great too though!

**Comment by robbiehanson** [comment]
Fixed via pull request #505
